docs/design/CONFIG_API.md

```diff
@@ -0,0 +1,319 @@
+
+// Provider defines the desired configuration of a provider.
+// +union
+type Provider struct {
+	// Type is the type of provider to use. If unset, the Kubernetes provider is used.
+	//
+	// +unionDiscriminator
+  Type ProviderType `json:"type,omitempty"`
+	// Kubernetes defines the configuration of the Kubernetes provider. Kubernetes
+	// provides runtime configuration via the Kubernetes API.
+	//
+	// +optional
+	Kubernetes *kubernetes.Provider `json:"kubernetes,omitempty"`
+
+	// File defines the configuration of the File provider. File provides runtime
+	// configuration defined by one or more files.
+	//
+	// +optional
+	File *file.Provider `json:"file,omitempty"`
+}
+
+// ProviderType defines the types of providers supported by Envoy Gateway.
+type ProviderType string
+
+const (
+	// KubernetesProviderType defines the "Kubernetes" provider.
+	KubernetesProviderType ProviderType = "Kubernetes"
+
+	// FileProviderType defines the "File" provider.
+	FileProviderType ProviderType = "File"
+)
+```
+Note that a provider-specific configuration is defined in the provider package. The following is an example of the
+Kubernetes provider:
+```go
+// gateway/pkg/kubernetes/kubernetes.go
+
+package kubernetes
+
+// Provider defines the configuration of the Kubernetes provider.
+type Provider struct {
+	// TODO: Define Kubernetes configuration fields, e.g. restrict namespaces to watch
+	//       Gateway/HTTPRoute resources.
+}
+```
+
+### Gateway
+Gateway defines desired configuration of [Gateway API][gw_api] controllers that reconcile and translate Gateway API
+resources into the Intermediate Representation (IR). Refer to the Envoy Gateway [design doc][design_doc] for additional
+details.
+
+### Provider
+Provider defines the desired configuration of an Envoy Gateway provider. A provider is an infrastructure component that
+Envoy Gateway calls to establish its runtime configuration. Provider is a [union type][union]. Therefore, Envoy Gateway
+can be configured with only one provider based on the `type` discriminator field. Refer to the Envoy Gateway
+[design doc][design_doc] for additional details.
+
+### Configuration Examples
+The configuration file is defined by the EnvoyGatewaySpec API type. At startup, Envoy Gateway searches for the
+configuration at "/etc/envoy-gateway/config.yaml".
+
+Start Envoy Gateway:
+```shell
+envoy-gateway start
+```
+Since the configuration file does not exist, Envoy Gateway will start with default configuration parameters.
+
+The Kubernetes provider can be configured explicitly using `provider.kubernetes`:
+```yaml
+$ cat << EOF > /etc/envoy-gateway/config.yaml
+provider:
+  type: Kubernetes
+  kubernetes: {}
+EOF
+```
+This configuration will cause Envoy Gateway to use the Kubernetes provider with default configuration parameters.
+
+The Kubernetes provider can be configured using the `provider` field. For example, the `foo` field can be set to "bar":
+```yaml
+$ cat << EOF > /etc/envoy-gateway/config.yaml
+provider:
```

Comment:
if we had used protobufs this would be looked a little cleaner
```
provider:
  kubernetes:
    ....
```

Reply:
Understood. As we discussed, `provider` follows the union type [conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions).